firecracker-microvm/firecracker
# [RFC] rfstool:  scripted rootfs builds.

## Reason for This PR

#1732 Outlines one possible approach to building rootfs filesystems.  

## Description of Changes

Adds the "rfstool" and associated Dockerfile for building a container image with wget and libguestfs-tools installed.   The tool is currently restricted to build just the minimal test rootfs.  It will fetch a minifs image from the Alpine repos, extract it, then customise the OS in order to make it suitable for running on Firecracker.

To test:

1. Build Docker image from the Dockerfile in the commit
2. Amend the line `RFSCTR_IMAGE=` in the rfstool script to point to your new docker image
3. Run `./tools/rfstool build` 

Expected output: 
```
$ ./tools/rfstool build
[Firecracker rfstool] Starting rootfs minimal build ...
[   0.0] Examining the guest ...
[  40.4] Setting a random seed
virt-customize: warning: random seed could not be set for this type of 
guest
[  40.9] Installing packages: ca-certificates
[  50.6] Installing packages: curl
[  62.1] Installing packages: iperf3
[  73.1] Installing packages: iproute2
[  87.4] Installing packages: iptables
[  99.4] Installing packages: openrc
[ 112.3] Installing packages: openssh-server
[ 125.7] Installing packages: util-linux
[ 148.8] Linking: /etc/init.d/agetty.ttyS0 -> agetty
[ 148.9] Running: echo ttyS0 >/etc/securetty
[ 149.9] Running: rc-update add agetty.ttyS0 default
[ 151.0] Running: rc-update add devfs boot
[ 152.2] Running: rc-update add procfs boot
[ 153.4] Running: rc-update add sysfs boot
[ 154.6] Running: rc-update add local default
[ 156.0] Writing: /etc/local.d/umount.stop
[ 156.1] Changing permissions of /etc/local.d/umount.stop to 0700
[ 156.3] Setting passwords
[ 170.1] Finishing off
[Firecracker rfstool] Minimal rootfs built and available at: ./firecracker/build/img/local/x86_64/minimal/fsfiles/boottime-rootfs.ext4
```

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
